File: src/coordination/AsyncTimerEvent.ts
// @leosingleton/commonlibs - Common Libraries for TypeScript and .NET Core
// Copyright (c) Leo C. Singleton IV <leo@leosingleton.com>
// See LICENSE in the project root for license information.

import { AsyncEventWaitHandle } from './AsyncEventWaitHandle';
import { IDisposable } from '../dotnet/Disposable';

/** Timer that behaves line an `EventWaitHandle`. Useful for the `AsyncEventWaitHandle.WhenAny()` method. */
export class AsyncTimerEvent extends AsyncEventWaitHandle implements IDisposable {
  public constructor(millisecondsDelay: number, repeat = false) {
    super(repeat, false);

    const f = repeat ? setInterval : setTimeout;
    this.intervalID = f(() => this.setEvent(), millisecondsDelay);
  }

  public dispose(): void {
    if (this.intervalID) {
      // According to https://developer.mozilla.org/en-US/docs/Web/API/WindowOrWorkerGlobalScope/clearInterval
      // clearInterval() and clearTimeout() are interchangeable and the intervalIDs are never reused. So it's safe to
      // simply call clearInterval() without tracking whether repeat == true or whether the non-repeatable timer has
      // ever fired.
      clearInterval(this.intervalID);
    }
  }

  private intervalID: number;
}

File: dotnet/CommonLibs/Coordination/AsyncTimerEvent.cs
// LeoSingleton.CommonLibs - Common Libraries for TypeScript and .NET Core
// Copyright (c) Leo C. Singleton IV <leo@leosingleton.com>
// See LICENSE in the project root for license information.

using System.Threading.Tasks;

namespace LeoSingleton.CommonLibs.Coordination
{
    /// <summary>
    /// Timer that behaves like an EventWaitHandle. Useful for the
    /// <see cref="AsyncEventWaitHandle.WhenAny(AsyncEventWaitHandle[])"/> method to set a timeout.
    /// </summary>
    public class AsyncTimerEvent : AsyncEventWaitHandle
    {
        /// <summary>
        /// Constructor
        /// </summary>
        /// <param name="millisecondsDelay">Timer interval, in milliseconds</param>
        /// <param name="repeat">If true, the timer repeats indefinitely; if false, it only occurs once.</param>
        public AsyncTimerEvent(int millisecondsDelay, bool repeat = false) : base(repeat, false)
        {
            var unused = Task.Run(async () =>
            {
                do
                {
                    await Task.Delay(millisecondsDelay).ConfigureAwait(false);
                    Set();
                } while (repeat);
            });
        }
    }
}
